jaxopt/_src/osqp.py

```diff
@@ -0,0 +1,474 @@
+    #     (2)        + (rho/2)   \|z_bar - z + rho^{-1} y\|^2_2
+    #     under    A x_bar = z_bar; x_bar = x
+    #
+    #        (1) and (2) come from the augmented Lagrangian
+    #
+    # This problem is easy to solve by writing the KKT optimality conditions.
+    # By construction the solution is unique without imposing strict convexity of objective nor 
+    # independance of the constraints.
+    # The primal feasability conditions are used to eliminate z_bar from the system (which simplifies it).
+    # The stationarity conditons obtained by derivating the Lagrangian wrt x_bar and z_bar
+    # constitute the two parts of this system.
+    x, z = params.primal
+    y    = params.dual_eq  # dual variables for constraints z_bar = z;
+    def matvec(unknowns):
+      x_bar, nu = unknowns
+      Ax_bar, AT_nu = _matvec_and_rmatvec(A, x_bar, nu)
+      stationarity_x_bar = tree_add(tree_add_scalar_mul(P(x_bar), self.sigma, x_bar), AT_nu)
+      stationarity_z_bar = tree_add(Ax_bar, tree_mul(tree_neg(inv_rho), nu))
+      return stationarity_x_bar, stationarity_z_bar
+    b1 = tree_sub(tree_scalar_mul(self.sigma, x), q)
+    b2 = tree_sub(z, tree_mul(inv_rho, y))
+    x_bar, nu = self.eq_qp_solve(matvec, (b1, b2), x0=eq_qp_last_sol,
+                                 maxiter=self.eq_qp_solve_maxiter,
+                                 tol=self.eq_qp_solve_tol)
+    return x_bar, nu
+
+  def _admm_step(self, params, P, q, A, box, rho, eq_qp_last_sol):
+    """Perform atomic step of ADMM algorithm."""
+    x, z = params.primal
+    y    = params.dual_eq  # dual variables for constraints z_bar = z;
+    # mu, phi = params.dual_ineq are unused
+
+    inv_rho = tree_map(lambda r: 1/r, rho)
+
+    # lines are numbered according to the pseudo-code in the paper OSQP.
+
+    # line 3: optimization step for (x_bar, z_bar)
+    # this equality constrained QP is solved by writing KKT conditions
+    # which fall back to a well posed linear system
+    x_bar, nu = self._solve_linear_system(params, P, q, A, inv_rho, eq_qp_last_sol)
+    z_bar = tree_add(z, tree_mul(inv_rho, tree_sub(nu, y)))  # line 4
+
+    # line 5: optimization step for x with relaxation parameter alpha (smooth updates)
+    # alpha can be understood as momentum term
+    x_next = tree_add(x, tree_scalar_mul(self.alpha, tree_sub(x_bar, x)))
+
+    # line 6: optimization step for z with relaxation parameter alpha (smooth updates)
+    # by definition A x_bar = z_bar and l <= z <= u thanks to projection
+    # the dual variable y corresponds to constraint z_bar = z
+    z_momentum = tree_add(z, tree_scalar_mul(self.alpha, tree_sub(z_bar, z)))
+    z_step = tree_mul(inv_rho, y)
+    z_free = tree_add(z_momentum, z_step)
+    z_next = projection_box(z_free, box)
+
+    # line 7: gradient descent on dual variables, with relaxation
+    y_step = tree_sub(z_momentum, z_next)
+    y_next = tree_add(y, tree_mul(rho, y_step))
+
+    return (x_next, z_next), y_next, (x_bar, nu)
```

Comment:
Could we define the implicit diff root function as the fixed point equations defined by ADMM? Could that help getting a better jacobian by leveraging the algorithm structure, rather than just having the KKT system?

Reply:
Yes; I found [this paper](https://arxiv.org/abs/1705.00772) the other day, and thought we might also be able to differentiate through the fixed points.

Reply:
It's true the box constraint projection isn't continuously differentiable though...